dash/_validate.py

```diff
@@ -0,0 +1,571 @@
+import collections
+import re
+
+from .development.base_component import Component
+from .dependencies import Input, Output, State, MATCH, ALLSMALLER
+from . import exceptions
+from ._utils import patch_collections_abc, _strings, stringify_id
+
+
+def validate_callback(app, layout, output, inputs, state):
+    is_multi = isinstance(output, (list, tuple))
+    validate_ids = not app.config.suppress_callback_exceptions
+
+    if layout is None and validate_ids:
+        # Without a layout, we can't do validation on the IDs and
+        # properties of the elements in the callback.
+        raise exceptions.LayoutIsNotDefined(
+            """
+            Attempting to assign a callback to the application but
+            the `layout` property has not been assigned.
+            Assign the `layout` property before assigning callbacks.
+            Alternatively, suppress this warning by setting
+            `suppress_callback_exceptions=True`
+            """
+        )
+
+    if not inputs:
+        raise exceptions.MissingInputsException(
+            """
+            This callback has no `Input` elements.
+            Without `Input` elements, this callback will never get called.
+
+            Subscribing to Input components will cause the
+            callback to be called whenever their values change.
+            """
+        )
+
+    outputs = output if is_multi else [output]
+
+    for args, cls in [(outputs, Output), (inputs, Input), (state, State)]:
+        validate_callback_args(args, cls, layout, validate_ids)
+
+    prevent_duplicate_outputs(app, outputs)
+    prevent_input_output_overlap(inputs, outputs)
+    prevent_inconsistent_wildcards(outputs, inputs, state)
```

Comment:
I've been wondering if it would make sense to move deeper graph validation to the renderer. If we later decide to allow multiple non-overlapping callbacks to update the same output for example, we would then only update the renderer, not every backend. The renderer could show the error in the DevTools and send back the error through a logging service to the backend to use. Might also be necessary or useful if/when we would come to support adding callbacks on the fly.